spccore/internal/cache.py

```diff
@@ -0,0 +1,235 @@
+            cache_map[file_path] = modified_time
+            _write_cache_map(cache_map, cache_dir)
+
+        return cache_map
+
+    def remove(self, file_handle_id: int, *, file_path: str = None, delete_file: bool = False) -> list:
+        """
+        Performs a cache Write to remove a file(s) from the cache.
+
+        :param file_handle_id: the file handle id to look up
+        :param file_path: the file_path to look up and remove.
+            Default None, which will remove all copies found in the cache.
+        :param delete_file: Set to True to remove the actual file. Default False.
+
+        :returns: A list of files removed
+        """
+        validate_type(int, file_handle_id, "file_handle_id")
+        removed = []
+        cache_dir = self.get_cache_dir(file_handle_id)
+
+        # to keep the critical section minimal, performs operations that does not need to lock prior to acquiring lock
+        file_path = normalize_path(file_path)
+        if delete_file and file_path and os.path.exists(file_path):
+            # if the actual file is deleted and we still keep a ref in the cache, that ref is out of date
+            # the cache will continue to work as expected
+            os.remove(file_path)
+            removed.append(file_path)
+
+        with Lock(SYNAPSE_DEFAULT_CACHE_MAP_FILE_NAME, current_working_directory=cache_dir):
+            cache_map = _get_cache_map(cache_dir)
+
+            if file_path is None:
+                for path in cache_map:
+                    if delete_file is True and os.path.exists(path):
+                        os.remove(path)
+                    removed.append(path)
+                cache_map = {}
+            elif file_path in cache_map:
+                del cache_map[file_path]
+
+            _write_cache_map(cache_map, cache_dir)
+
+        return removed
+
+    def purge(self, before_date: datetime.datetime, *, dry_run: bool = False) -> int:
+        """
+        Purge the cache. Use with caution. Delete files whose cache maps were last updated prior to the given date.
+        Deletes .cacheMap files and files that the cache map point to.
+
+        This function is recommended when working with PHI or sensitive data.
+
+        :param before_date: the cutoff time to look up
+        :param dry_run: Set to True to list all cache dir that would be removed. Default False.
+        :return: the number of file handle id directories that has been removed
+        """
+        validate_type(datetime.datetime, before_date, "before_date")
+        before_date = from_datetime_to_epoch_time(before_date)
+        count = 0
+        for cache_dir in _cache_dirs(self.cache_root_dir):
+            # _get_modified_time returns None if the cache map file doesn't exist.
+            # We are going to purge directories in the cache that have no .cacheMap file.
+            last_modified_time = get_modified_time(os.path.join(cache_dir, SYNAPSE_DEFAULT_CACHE_MAP_FILE_NAME))
```

Comment:
need to loop through each entry in the .cacheMap file and check its modifiedOn timestamp